ddtrace/_monkey.py

```diff
@@ -178,6 +180,57 @@ class ModuleNotFoundException(PatchException):
     pass
 
 
+def is_version_compatible(package_name: str, version: str, supported_versions: Dict[str, str]) -> bool:
+    """Check if a given package version is compatible with the instrumented versions.
+
+    Args:
+        package_name: The name of the package to check
+        version: The version string to check against
+        supported_versions: Dictionary mapping module names to their version requirements
+
+    Returns:
+        bool: True if the version is compatible, False otherwise
+    """
+    if package_name not in supported_versions:
+        return False
+
+    spec = supported_versions.get(package_name)
+    if not spec:
+        return False
+
+    version_specifier = VersionSpecifier(spec)
+    return version_specifier.contains(version)
+
+
+def _get_installed_version(imported_module, module):
+    if hasattr(imported_module, "get_version"):
+        return imported_module.get_version()
+    elif hasattr(imported_module, "get_versions"):
+        return imported_module.get_versions().get(module)
+    return None
+
+
+def should_patch_module(imported_module, module):
+    if not hasattr(imported_module, "_supported_versions"):
+        return True
+
+    supported_versions = imported_module._supported_versions()
+
+    installed_version = _get_installed_version(imported_module, module)
+
+    if not installed_version:
+        return True
+
+    if not is_version_compatible(module, installed_version, supported_versions):
+        message = (
+            f"Skipped patching '{module}' integration, installed version: {installed_version} is not compatible with "
+            f"integration support spec: {supported_versions[module]}"
+        )
+        log.debug(message)
```

Comment:
```suggestion
        raise IncompatibleModuleException(message)
```

Will need this defined above:

```python
class IncompatibleModuleException(PatchException):
    pass
```


Then it'll hit the error logging and telemetry log failing to patch for this integration, and it'll simplify some of the calling code too.

```python
if config._trace_safe_instrumentation_enabled:
    should_patch_module(imported_module, module). # maybe a name more like "check_module_compatible" might be better if the function returns None and relies on raising an exception... 🤷🏻 
imported_module.patch()
if hasattr(...):
    ...
```


Not sure this is actually better, but could make the `_on_import_factory` changes only be 2 lines only ?


Depends on whether you actually want these to error log or not though.

Reply:
Logging the error is fine, just don't want to actually throw the error so I added a check to prevent that. I've incorporated these changes 👍